src/hotspot/share/opto/library_call.cpp

```diff
@@ -1046,18 +1046,20 @@ bool LibraryCallKit::inline_preconditions_checkIndex(BasicType bt) {
   if (!rc_bool->is_Con()) {
     record_for_igvn(rc);
   }
+
+  if (stopped()) {
+    return false;
+  }
+
   set_control(_gvn.transform(new IfTrueNode(rc)));
+
   {
     PreserveJVMState pjvms(this);
     set_control(_gvn.transform(new IfFalseNode(rc)));
     uncommon_trap(Deoptimization::Reason_range_check,
                   Deoptimization::Action_make_not_entrant);
   }
 
-  if (stopped()) {
```

Comment:
Moving the check around doesn't make much sense to me.

`stopped() == false` signals that the current control is effectively dead. It could happen when the range check (`RangeCheckNode` )  always fails and execution unconditionally hits the uncommon trap. (I haven't double-checked myself whether it happens in practice or not.) 

By bailing out from the intrinsic (`return false;`), the next thing C2 will attempt is to inline `Preconditions::checkIndex(). There were no attempts to clean up the graph built up to this point (with the uncommon trap).

Instead, the fix can just be `if (stopped()) { return true; }`. The graph constructed so far is valid.

Reply:
Yes, returning `true` could be correct because of constant folding we end up in uncommon trap which is valid for provided values (constants). It will be the same if code is compiled normally.
But we also need to move the other `stopped()` check from line 1036 to 1028 after BuildCutout which may also can go to uncommon trap (when length is know negative during compilation). And it also should return `true`.